osquery/core/database/rocksdb_database.cpp

```diff
@@ -0,0 +1,388 @@
+      assert(false && type_error.getFullMessageRecursive().c_str());
+      return createError(DatabaseError::KeyNotFound, "", std::move(type_error));
+    }
+#endif
+    return result_str;
+  }
+  return result.takeError();
+}
+
+ExpectedSuccess<DatabaseError> RocksdbDatabase::putString(
+    const std::string& domain,
+    const std::string& key,
+    const std::string& value) {
+  auto result = putRawBytes(domain, key, value);
+  if (result) {
+    return Success();
+  }
+  return result.takeError();
+}
+
+Expected<int, DatabaseError> RocksdbDatabase::getInt32(
+    const std::string& domain, const std::string& key) {
+  Expected<std::string, DatabaseError> buffer = getRawBytes(domain, key);
+  if (buffer) {
+    std::string value = buffer.take();
+    if (BOOST_LIKELY(value.back() == kIntTag)) {
+      int result = *(reinterpret_cast<const int*>(value.data()));
+      return ntohl(result);
+    } else {
+      auto type_error = createError(RocksdbError::UnexpectedValueType,
+                                    "Fetching string as integer");
+      auto error =
+          createError(DatabaseError::KeyNotFound, "", std::move(type_error));
+      assert(false && error.getFullMessageRecursive().c_str());
+      LOG(ERROR) << error.getFullMessageRecursive();
+      debug_only::fail(error.getFullMessageRecursive().c_str());
+      return std::move(error);
+    }
+  }
+  return buffer.takeError();
+}
+
+ExpectedSuccess<DatabaseError> RocksdbDatabase::putInt32(
+    const std::string& domain, const std::string& key, const int32_t value) {
+  int tmp_value = htonl(value);
+  std::string buffer;
+  buffer.reserve(sizeof(int32_t) + sizeof(int8_t));
+  buffer.append(reinterpret_cast<const char*>(&tmp_value), 4);
+  buffer.append(reinterpret_cast<const char*>(&kIntTag), 1);
+  return putString(domain, key, buffer);
+}
+
+ExpectedSuccess<DatabaseError> RocksdbDatabase::putStringsUnsafe(
+    const std::string& domain,
+    std::vector<std::pair<std::string, std::string>>& data) {
+  auto handle = getHandle(domain);
+  if (handle) {
+    std::shared_ptr<Handle> handle_ptr = handle.take();
+    rocksdb::WriteBatch batch;
+
+    for (const auto& pair : data) {
+      auto status = batch.Put(handle_ptr.get(), pair.first, pair.second);
+      if (!status.ok()) {
+        auto batch_write_error =
+            createError(RocksdbError::BatchWriteFail, status.ToString());
+        return createError(DatabaseError::FailToWriteData,
+                           "Failed to write data",
+                           std::move(batch_write_error));
+      }
+    }
+    auto status = db_->Write(batch_write_options_, &batch);
+    if (!status.ok()) {
+      return createError(DatabaseError::FailToWriteData, status.ToString());
+    }
+  }
+  return handle.takeError();
+}
+
+Expected<std::vector<std::string>, DatabaseError> RocksdbDatabase::getKeys(
+    const std::string& domain, const std::string& prefix) {
+  auto handle = getHandle(domain);
```

Comment:
It is not exactly a `handle`, is a Expected handle. That make construction `...HandleRef handle_ptr = handle.take();` looks a bit surprising 😯. Would you like to give more talkative name to it?